flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,343 @@
+            idType =
+                    LogicalTypeDataTypeConverter.toDataType(
+                            LogicalTypeParser.parse((String) this.meta.get("idType")));
+            modelData.setIdType(idType);
+        }
+    }
+
+    /**
+     * this operator use mapper to load the model data and do the prediction. if you want to write a
+     * prediction operator, you need implement a special mapper for this operator.
+     */
+    private static class PredictOperator
+            extends AbstractUdfStreamOperator<Row, RichMapFunction<Row, Row>>
+            implements OneInputStreamOperator<Row, Row> {
+
+        public PredictOperator(RichMapFunction<Row, Row> userFunction) {
+            super(userFunction);
+        }
+
+        @Override
+        public void processElement(StreamRecord<Row> streamRecord) throws Exception {
+            Row value = streamRecord.getValue();
+            output.collect(new StreamRecord<>(userFunction.map(value)));
+        }
+    }
+
+    public ResolvedSchema getOutputSchema(
+            ResolvedSchema dataSchema, Map<Param<?>, Object> params, DataType idType) {
+        String[] reservedCols = (String[]) params.get(KnnParams.RESERVED_COLS);
+        reservedCols =
+                (reservedCols == null)
+                        ? dataSchema.getColumnNames().toArray(new String[0])
+                        : reservedCols;
+        DataType[] reservedTypes = findColTypes(dataSchema, reservedCols);
+        String[] resultCols = new String[] {(String) params.get(KnnParams.PREDICTION_COL)};
+        DataType[] resultTypes = new DataType[] {idType};
+        return ResolvedSchema.physical(
+                ArrayUtils.addAll(reservedCols, resultCols),
+                ArrayUtils.addAll(reservedTypes, resultTypes));
+    }
+
+    /** @return parameters for algorithm. */
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        if (null == this.params) {
+            this.params = new HashMap<>(1);
+        }
+        return this.params;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) modelData[0]).getTableEnvironment();
+
+        String dataPath = ReadWriteUtils.getDataPath(path);
+        FileSink<Row> sink =
+                FileSink.forRowFormat(new Path(dataPath), new KnnModelData.ModelDataEncoder())
+                        .withRollingPolicy(OnCheckpointRollingPolicy.build())
+                        .withBucketAssigner(new BasePathBucketAssigner<>())
+                        .build();
+        tEnv.toDataStream(modelData[0]).sinkTo(sink);
+        HashMap<String, String> meta = new HashMap<>(1);
+        meta.put("idType", modelData[0].getResolvedSchema().getColumnDataTypes().get(3).toString());
+        ReadWriteUtils.saveMetadata(this, path, meta);
+    }
+
+    public void load(StreamExecutionEnvironment env, String path) throws IOException {
```

Comment:
this method should be static.

And the corresponding `Estimator` should also have a static `load` method.

Reply:
Estimator just train the model. loading model for what?.

Reply:
this method should be static. I accept.
 I will change it later.

Reply:
done, thanks for your suggestion.

Reply:
Knn has some metadata, like `K`, that needs to be saved to file and reloaded to create an `Estimator`. You can refer to `Kmeans.load()` for implementation.

Even if we can find an `Estimator` that does not have any parameters, it should also implement a `load` method because that is how `Pipeline` is reconstructed. Pipelines have no idea whether a stage has to be loaded from file system or not, and its choice is to invoke the `load` method on each stage. This is also documented in `Stage`'s javadoc.